Sup 7 customer create/edit

alexeyoverchenko opened the pull request:
fix after review

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
Merge branch 'develop' into sup-7_customer_create/edit

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
Merge branch 'develop' into sup-7_customer_create/edit

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
bugFix in Customer

alexeyoverchenko committed:
create "get warehouse info" functional

alexeyoverchenko committed:
for review after fix

alexeyoverchenko committed:
for review after fix

alexeyoverchenko committed:
for review after fix

alexeyoverchenko committed:
for review after fix

alexeyoverchenko committed:
for review after fix

IrynaZazyba commented on src/main/java/by/itech/lab/supplier/controller/CustomerController.java:
```diff
@@ -26,15 +35,40 @@
+        customerService.activate(id);
+    }
+
+
```
extra line

IrynaZazyba commented on src/main/java/by/itech/lab/supplier/repository/CustomerRepository.java:
```diff
@@ -12,5 +13,12 @@
 
-}
+    @Modifying
+    @Query("update Customer set status = false where id = :id")
```
need discuss

svareyko replied:
whats the result of the discussion?

alexeyoverchenko replied:
I suggest renaming "activate" method in BaseServise to "switchStatus" 
and implement method like this

    public void switchStatus(final Long id) {
        final Customer existing = customerRepository
                .findById(id)
                .orElseThrow();
        if (existing.isStatus()) {
            customerRepository.deactivate(id);
        } else {
            customerRepository.activate(id);
        }

svareyko replied:
![image](https://user-images.githubusercontent.com/44436165/97689713-0cb84780-1aad-11eb-9be6-9d118ae99ebd.png)

svareyko commented on src/main/java/by/itech/lab/supplier/repository/UserRepository.java:
```diff
@@ -16,7 +16,7 @@
 
     @Modifying
-    @Query("update User set activeForDeletion = false where id = :id")
+    @Query("update User set deleted = false where id = :id")
```
from the logic side: if `deleted == true` - its not deleted, but if `false` - its deleted

svareyko replied:
is it ok?

alexeyoverchenko replied:
yes it will be better

svareyko replied:
are you going to do something with this?

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/CustomerServiceImpl.java:
```diff
@@ -9,24 +12,70 @@
+        return customerMapper.map(saved);
+    }
+
+    private UserDto createAdmin(CustomerDto customerDto) {
```
what's the correct method order?

svareyko replied:
maybe better to put this logic to UserService?as here you exactly doing a job which related only to that service

alexeyoverchenko replied:
Maybe, we discuss it tomorrow!

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/CustomerServiceImpl.java:
```diff
@@ -9,24 +12,70 @@
+                .build();
+    }
+
+    private void updateCustomer(final CustomerDto from, final Customer to) {
```
why it's not in the mapper?

alexeyoverchenko replied:
I've corrected it

svareyko reviewed (changes requested):
please fix comments

alexeyoverchenko committed:
get Warehouse info

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/CustomerServiceImpl.java:
```diff
@@ -9,24 +12,76 @@
+        final Customer existing = customerRepository
+                .findById(id)
+                .orElseThrow();
+        if (existing.isStatus()) {
```
![image](https://user-images.githubusercontent.com/44436165/97690087-8cdead00-1aad-11eb-9329-83f33b53360f.png)

svareyko replied:
for what you need this check? why you have added the unnecessary additional request to DB?

svareyko commented on src/main/java/by/itech/lab/supplier/service/impl/CustomerServiceImpl.java:
```diff
@@ -9,24 +12,76 @@
+                .findById(id)
+                .orElseThrow();
+        if (existing.isStatus()) {
+            customerRepository.deactivate(id);
```
we discussed yesterday that we don't need such a split, we can use only one repo method

alexeyoverchenko committed:
- rewrited and renamed method for change the activation. Now it's called "changeActiveStatus". and it's change status through one method
- changed @Query("update User set deleted = false where id = :id"). Now if customer was delete it looks like "deleted=true"
- I put createAdmin method in UserServiceImpl
- add createAdmin() in UserService
- I put updateCustomer() in mapper

- Also we made decision delete "activate" method from BaseService, and add changeActivateStatus in our Services if it needed. So I add changeActiveStatus in CustomerService

svareyko reviewed (approved):
In the UserRepository method which used for soft-delete one comment, fix it and lets merge

svareyko commented on src/main/java/by/itech/lab/supplier/controller/CustomerController.java:
```diff
@@ -59,10 +59,10 @@ public CustomerDto updateCustomer(@PathVariable Long id, @Valid @RequestBody Cus
-    public void activate(@PathVariable Long id) {
-        customerService.switchStatus(id);
+    @PutMapping(ApiConstants.URL_ID_PARAMETER + ApiConstants.URL_STATUS_PARAMETER)
+    @ResponseStatus(HttpStatus.OK)
```
what is the default status?

svareyko commented on src/main/java/by/itech/lab/supplier/controller/CustomerController.java:
```diff
@@ -26,15 +35,39 @@
+    }
+
+    @PutMapping(ApiConstants.URL_ID_PARAMETER)
+    @ResponseStatus(HttpStatus.OK)
```
what is the default status?

svareyko commented on src/main/java/by/itech/lab/supplier/service/CustomerService.java:
```diff
@@ -1,14 +1,10 @@
-    Page<CustomerDto> getCustomers(Pageable pageable, Boolean status);
-
-    CustomerDto getCustomer(Long customerId);
+    boolean changeActiveStatus(Long id, boolean status);
```
why it is still called status?

alexeyoverchenko committed:
-created change set for customer table (renamed "status" to "active" and added deleted_at column)
- added corresponding changes into the code

IrynaZazyba commented on src/main/java/by/itech/lab/supplier/dto/CustomerDto.java:
```diff
@@ -2,18 +2,21 @@
+    @Pattern(regexp = "[*?=%:]", message = "You have typed the wrong characters")
     private String name;
     private LocalDate registrationDate;
     private boolean status;
```
active?

IrynaZazyba commented on src/main/java/by/itech/lab/supplier/dto/mapper/CustomerMapper.java:
```diff
@@ -20,7 +19,7 @@ public Customer map(CustomerDto dto) {
-                .registrationDate(Date.valueOf(dto.getRegistrationDate()))
+                .registrationDate(dto.getRegistrationDate())
                 .status(dto.isStatus())
                 .build();
```
status? you change to active only in entity, correct dto and mapper

IrynaZazyba commented on src/main/resources/db/changelog/1.0/2020-11-02-change-table-customer.yaml:
```diff
@@ -0,0 +1,22 @@
+        - renameColumn:
+            newColumnName:  active
+            oldColumnName:  status
+            tableName:  customer
```
extra line

alexeyoverchenko committed:
- edits in dto and map
- edits in findall

alexeyoverchenko committed:
- last bugfix

alexeyoverchenko committed:
- add EqualsAndHashCode annotation

alexeyoverchenko committed:
- edit soft delete method
- set @transactional annotation for "save"
- restored LocalDate creation at the time customer creation

alexeyoverchenko committed:
- edit soft delete method
- set @transactional annotation for "save"
- restored LocalDate creation at the time customer creation
- redacted customer entity to CustomerDto in UserDto

alexeyoverchenko committed:
- edit soft delete method
- set @transactional annotation for "save"
- restored LocalDate creation at the time customer creation
- redacted customer entity to CustomerDto in UserDto
- edit connection entity with db in Address
- checked methods' functionality in CustomerController
- add todo annotations

alexeyoverchenko committed:
- removed todo
- made changeset (removed "not" null for address_id in User table)

alexeyoverchenko committed:
- removed todo
- made changeset (removed "not" null for address_id in User table)
- made changeset (removed "not" null for birthday in User table)

svareyko commented on src/main/resources/db/changelog/1.0/2020-11-02-change-table-customer.yaml:
```diff
@@ -0,0 +1,23 @@
+            columnDataType: boolean
+            newColumnName: active
+            oldColumnName: status
+            tableName: customer
```
empty line at the end

svareyko commented on src/main/resources/db/changelog/1.0/2020-11-04-change-table-user-2.yaml:
```diff
@@ -0,0 +1,11 @@
+            columnDataType: bigint
+            columnName:  birthday
+            schemaName:  test
+            tableName:  user
```
empty line at the end

svareyko commented on src/main/resources/db/changelog/1.0/2020-11-04-cnahge-table-user.yaml:
```diff
@@ -0,0 +1,11 @@
+            columnDataType: bigint
+            columnName:  address_id
+            schemaName:  test
+            tableName:  user
```
empty line at the end

alexeyoverchenko committed:
-add empty line in changesets
